### PR TITLE
Allow requesting an emojis creator.

### PR DESCRIPTION
**Pull request by Saladoc**
As seen on ~~TV~~ #364 and #371

### CODE REVIEW

File: javacord-api/src/main/java/org/javacord/api/entity/emoji/KnownCustomEmoji.java [hunk 1]
```diff
@@ -69,6 +70,13 @@ default CustomEmojiUpdater createUpdater() {
      */
     boolean isManaged();
 
+    /**
+     * Gets the creator of the emoji.
+     *
+     * @return The user who created the emoji.
+     */
+    CompletableFuture<Optional<User>> getCreator();
```
Comment (Vampire) [inline javacord-api/src/main/java/org/javacord/api/entity/emoji/KnownCustomEmoji.java]: Can there really be a custom emoji without a creator?

Comment (Bastian) [reply javacord-api/src/main/java/org/javacord/api/entity/emoji/KnownCustomEmoji.java]: I guess the field might not be present when you don't have enough permissions.
At least in the UI you can only view the creator of an emoji in the server settings:

File: javacord-core/src/main/java/org/javacord/core/entity/emoji/KnownCustomEmojiImpl.java [hunk 2]
```diff
@@ -111,6 +118,38 @@ public boolean isManaged() {
...
     }
 
+    @Override
+    public CompletableFuture<Optional<User>> getCreator() {
+        if (creatorId == 0L) {
+            CompletableFuture<Optional<User>> future = new CompletableFuture<>();
+            new RestRequest<Long>(getApi(), RestMethod.GET, RestEndpoint.CUSTOM_EMOJI)
+                    .setUrlParameters(server.getIdAsString(), this.getIdAsString())
+                    .execute(result -> result.getJsonBody().has("user")
+                            ? result.getJsonBody().get("user").asLong(0)
```
Comment (Bastian) [inline javacord-core/src/main/java/org/javacord/core/entity/emoji/KnownCustomEmojiImpl.java]: The endpoint returns a user entity, not a user id.
A response might look like this:
```json
{
   "managed":false,
   "name":"JC3",
   "roles":[

   ],
   "user":{
      "username":"Bastian",
      "discriminator":"0001",
      "id":"157862224206102529",
      "avatar":"a_796bef60840438a232f53f24ba29165c"
   },
   "require_colons":true,
   "animated":false,
   "id":"429927176268152842"
}
```

File: javacord-api/src/main/java/org/javacord/api/entity/emoji/KnownCustomEmoji.java (hunk 1 above)
Comment (Bastian) [reply javacord-api/src/main/java/org/javacord/api/entity/emoji/KnownCustomEmoji.java]: {"managed": false, "name": "JC3", "roles": [], "require_colons": true, "animated": false, "id": "429927176268152842"}

Comment (Bastian) [reply javacord-api/src/main/java/org/javacord/api/entity/emoji/KnownCustomEmoji.java]: Just tested it.
With permissions:
```json
{"managed": false, "name": "JC3", "roles": [], "user": {"username": "Bastian", "discriminator": "0001", "id": "157862224206102529", "avatar": "a_796bef60840438a232f53f24ba29165c"}, "require_colons": true, "animated": false, "id": "429927176268152842"}
```
Without permissions:
```json
{"managed": false, "name": "JC3", "roles": [], "require_colons": true, "animated": false, "id": "429927176268152842"}
```

File: javacord-core/src/main/java/org/javacord/core/entity/emoji/KnownCustomEmojiImpl.java [hunk 3]
```diff
@@ -111,6 +119,38 @@ public boolean isManaged() {
...
+                        }
+                        return null;
+                    })
+                    .exceptionally(t -> {
+                        future.completeExceptionally(t);
+                        return null;
+                    });
+            return future;
+        } else {
+            return getApi().getUserById(creatorId).thenApply(Optional::of);
```
[the remark this answers is not in the document]
Comment (Saladoc) [reply javacord-core/src/main/java/org/javacord/core/entity/emoji/KnownCustomEmojiImpl.java]: It's `getUserById`, not `getCachedUserById`. The creator is always requested if an id is stored and may be served from cache if it is present there. The user will have to store the creator as long as they deem it relevant. I see no point in having the creators of every emoji that were requested once clutter the cache if they are not relevant for other reasons.

[the remark this answers is not in the document]
Comment (Bastian) [reply javacord-core/src/main/java/org/javacord/core/entity/emoji/KnownCustomEmojiImpl.java]: My bad

File: javacord-core/src/main/java/org/javacord/core/entity/emoji/KnownCustomEmojiImpl.java [hunk 4]
```diff
@@ -60,6 +66,7 @@ public KnownCustomEmojiImpl(DiscordApiImpl api, Server server, JsonNode data) {
             }
         }
         requiresColons = !data.hasNonNull("require_colons") || data.get("require_colons").asBoolean();
+        creatorId = data.has("creator") ? data.get("creator").asLong() : 0L;
```
Comment (Bastian) [inline javacord-core/src/main/java/org/javacord/core/entity/emoji/KnownCustomEmojiImpl.java]: Why `creator` and not `user`?

**Comment by Bastian** [comment]
There's still one open question I would like to answer: What happens when the creator of the emoji deletes his account? I can imagine that `getUserById(...)` will fail with a 404 Not Found, but the `user` field may still contain the old name and id of the user.

File: javacord-core/src/main/java/org/javacord/core/entity/emoji/KnownCustomEmojiImpl.java (hunk 4 above)
Comment (Saladoc) [reply javacord-core/src/main/java/org/javacord/core/entity/emoji/KnownCustomEmojiImpl.java]: missed that one when I corrected the rest.

**Comment by Bastian** [comment]
As I have no clue how long it may take for account to be deleted, I'm going to merge the current state of the PR.
In case account deletion is causing issues, we can still fix it.

**Pull request merged**